### PR TITLE
Space raider zhechao

### PR DESCRIPTION
**Pull request by zhechao**
Created the black hole object and loop-hole object

### CODE REVIEW

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: All the encounter game object should extend the `EncounterObject` (_objects/pattern/encounterObject.cs_). Please refer to this class, you might find that a lot of useful things are already written ;)

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: This is a little bit of nit-picking, but you are missing space between the word _Constructs_ and an article _an_. Please consider rewording this.

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: Our part of the game (and I think the whole game) will rely on vectors (instead of separate coordinates). Please consider rewriting this so it uses the class `Vector` (_utils/vector.js_).

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: I think you meant to use `this.bhImage.src` instead of `this.bhImage`, but I do not think this code works as intended. Are you sure you tested this properly? o.O 

Also, we have a utility called `AssetLoader` (_utils/assetLoader.cs_), which ensures loading and caching of assets. Please consider using this utility.

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js:48]: I tried to understand this piece of code (also made three people look at it) but I cannot wrap my mind around it. Would you please mind explaining it or adding some comments/submethods? Thank you.

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: Once again, please consider using the `superUpdate` method of `EncounterObject`.

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: First of all, our game engine does not involve mass. If you think it should, please consider adding it _yourself_ to all related game objects. 

Second, I have my serious doubts about you actually trying those things out. There are several typos which make this code useless (e.g. `shipY` x `Shipy`). Please consider checking this file again.

Next, what do those ridiculous constants stand for? Could you please encapsulate them into separate method and comment, why have you decided to choose them?

Last, of all, all the collisions should be handled in the dedicated class called `CollisionHandler` (_objects/pattern/collisionHandler.js_). Please refer to the cases used there to implement this kind of behavior.

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: I have mentioned a few times already, but please use Vector library.

**Review by ksu-shanemat — CHANGES_REQUESTED** [review]
You have pushed in a lot of what seems to be VS configuration files ...

File: src/space/objects/realization/blackhole.js
Comment (zhechao) [reply src/space/objects/realization/blackhole.js]: in the game, I think a black hole won't move at all, the same as the loop hole. If I use vector, it will be {x:0, y:0}.
OR you want to me make it as a moveable black hole?

File: src/space/objects/realization/blackhole.js
Comment (zhechao) [reply src/space/objects/realization/blackhole.js]: fixed, but not push the update yet

File: src/space/objects/realization/blackhole.js
Comment (zhechao) [reply src/space/objects/realization/blackhole.js:48]: I assume the position of black hole in the map won't be changed, so I just use this update to make the black hole sprite work.

Comment (zhechao) [reply src/space/objects/realization/blackhole.js:48]: I will add comment there once I finished the asteroids part

File: src/space/objects/realization/blackhole.js
Comment (zhechao) [reply src/space/objects/realization/blackhole.js]: For the first question, I asked in group talk, someone told me I should have ship mass
Second, I need the ship's coordinate to calculate the distance between the center of black hole and ship.
Third, there is no collision between black hole and ship, the only thing black hole will do is pull the ship.

Comment (ksu-shanemat) [reply src/space/objects/realization/blackhole.js]: Still, you have not responded to my suggestion. Of course, adding mass to _all encounter objects_ might be a good idea, but if you are the only one actually needing that, then you should add such a change _yourself_.

Second, there should be a reference to the player ship accessible from the base _Encounter_ object, try looking for that one.

Third, yes, but as we discussed in class, this should be treated as a "collision" with the black hole area.

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [reply src/space/objects/realization/blackhole.js:48]: But this function relies on the fixed position of [500, 100] and [0, 0]. This should (at least imo) be made more general.

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [reply src/space/objects/realization/blackhole.js]: It doesn't really matter, whether the black hole will move or not - the whole game is using vectors. If you don't want the black hole to move, do not set it's velocity, since it defaults to (0, 0). 

Whether the black hole will move or not, that decision I will leave to you ;)

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js:22]: As I mentioned in my previous comment - could you please consider using our utility class called _AssetLoader_? I am afraid that if all the asset loading is not in one place, we might forget this change while trying to merge our group branch with others. Thank you

**Comment by ksu-shanemat** [comment]
I would like to state, that you should not merge any PR that has been marked as _changes requested_. I believe you have done this unintentionally. I will remove the merge commit from our master branch and ask you, to first resolve any comments any of the developers have before merging it. Thank you

**Comment by zhechao** [comment]
I have deleted what I merged, and just pushed a new request

**Pull request merged**
